Add token healing to `main` and `server`

mare5x opened the pull request:
Add different token healing strategies to `main` and `server`. Token healing works by chopping off some tokens from the tokenized prompt and then constraining the decoding to match the bytes of the removed tokens.

Example usage:
```bash
# with token healing
$ ./llama-cli -m ./models/phi-2/ggml-model-q4_0.gguf --temp 0 -p "print('Hello, wo" -th d
print('Hello, world!')

# without token healing
$ ./llama-cli -m ./models/phi-2/ggml-model-q4_0.gguf --temp 0 -p "print('Hello, wo" -th 0
print('Hello, woof!')
```

Server usage:
```bash
prompt='print(\"Hello, Wo'

curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '
    {
      "prompt": '\""${prompt}"\"',
      "n_predict": 10,
      "temperature": 0
    }
' | jq '. | {prompt, content}'

curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '
    {
      "prompt": '\""${prompt}"\"',
      "n_predict": 10,
      "temperature": 0,
      "token_healing": "d"
    }
' | jq '. | {prompt, content}'
```

Solves https://github.com/ggerganov/llama.cpp/issues/5765.

ggerganov reviewed (approved):
Looks OK, though I haven't ran any tests. Would appreciate if people play with this and provide some feedback before merging

Overall, I'm a bit concerned that the logic for managing the sampling state and the KV cache in the examples starts to become very convoluted when we add such kind of techniques (e.g. context shifting, self-extend, templates, speculation, ...). We have to start refactoring this and simplify it.

It's not specific to this PR, just a general thought that I have lately. Will try to prioritise this a bit, though I don't have a very good idea what specifically needs to be done yet, so any ideas and help from others of how to simplify the code are appreciated

mare5x committed:
server : add token healing support

mare5x committed:
main : add token healing

mare5x committed:
token healing : change dynamic rollback

Dynamic rollback now starts checking prefixes based on the length of the longest token.

mare5x committed:
token healing : refactor to return struct

mare5x committed:
token healing : handle more special tokens

Infill tokens were being rolled back in certain cases.

mare5x committed:
server : token healing for infilling/FIM

mare5x committed:
token healing : refactor argument parsing

Unify `main` and `server` token healing argument handling.

mare5x committed:
token healing : change argument order

shibe2 commented on examples/server/README.md:
```diff
@@ -266,6 +266,8 @@ node index.js
 
     `json_schema`: Set a JSON schema for grammar-based sampling (e.g. `{"items": {"type": "string"}, "minItems": 10, "maxItems": 100}` of a list of strings, or `{}` for any JSON). See [tests](../../tests/test-json-schema-to-grammar.cpp) for supported features.  Default: no JSON schema.
 
+    `token_healing`: Set token healing strategy. Default: `0`, which is disabled.
```
Maybe add possible values for `token_healing` to this document too. And/or a link to where this is explained.

mare5x committed:
readme : list possible token healing values

mare5x committed:
token healing : fix rebase bug

mare5x commented:
I am sharing my results of evaluating token healing with a simple [evaluation script](https://github.com/mare5x/llama.cpp/blob/token-healing-eval/scripts/evaluate_repoeval.py) that uses either `llama-cli` or the server's `/completion` endpoint.
- model: [tiny_starcoder_py.Q8_0.gguf](https://huggingface.co/RichardErkhov/bigcode_-_tiny_starcoder_py-gguf) (164M params)
- dataset: [RepoEval](https://github.com/microsoft/CodeT/tree/main/RepoCoder)
  - `line`: line-level RepoEval dataset (predict one full line of code), 1k context size
  - `lineR`: pass `--randomize_target` to the eval script (start completing from a random position in the target line) 
- generation args: `-c 1536 --temp 0 -n 25` for cli and `'{"temperature": 0, "n_predict": 25, "stop": ["\n"]}'` for server (+ token healing option)
- metrics: exact match (EM); edit similarity (ES)

|        | token healing | `line` EM | `line` ES | `lineR` EM | `lineR` ES |
|--------|---------------|-----------|-----------|------------|------------|
| server | none (0)      | 2.81      | 24.18     | 21.31      | 48.75      |
| server | 1             | 25.06     | 51.25     | 41.19      | 63.88      |
| server | d1            | 25.06     | 51.25     | 46.00      | 66.78      |
| server | r3            | 24.56     | 50.98     | 48.00      | 69.01      |
| server | d             | 25.06     | 51.25     | 48.25      | 69.17      |
| cli    | none (0)      | 3.62      | 26.46     | 20.06      | 48.41      |
| cli    | d             | 25.06     | 52.02     | 45.56      | 67.88      |

- With `line`, all token healing options simply roll back the last `\n` which is why the token healing results are similar.
- `d` > `r3` > `d1` > `1`
